src/storage/mvcc/txn.rs

```diff
@@ -485,8 +485,34 @@ impl<S: Snapshot> MvccTxn<S> {
     }
 
     pub fn rollback(&mut self, key: Key) -> Result<bool> {
+        self.cleanup(key, 0)
+    }
+
+    /// Cleanup the lock if it's TTL has expired, comparing with `current_ts`. If `current_ts` is 0,
+    /// cleanup the lock anyway.
+    ///
+    /// Returns whether the lock is a pessimistic lock. Returns error if the key has already been
+    /// committed.
+    pub fn cleanup(&mut self, key: Key, current_ts: u64) -> Result<bool> {
         let is_pessimistic_txn = match self.reader.load_lock(&key)? {
-            Some(ref lock) if lock.ts == self.start_ts => {
+            Some(ref mut lock) if lock.ts == self.start_ts => {
+                // If current_ts is not 0, check the Lock's TTL.
+                // If the lock is not expired, do not rollback it but report key is locked.
+                if current_ts > 0
+                    && extract_physical(lock.ts) + lock.ttl >= extract_physical(current_ts)
+                {
+                    // The `lock.primary` field will not be accessed again. Use mem::replace to
+                    // avoid cloning.
+                    let primary = ::std::mem::replace(&mut lock.primary, Default::default());
+                    let mut info = kvproto::kvrpcpb::LockInfo::default();
+                    info.set_primary_lock(primary);
```

Comment:
I think it's already moved here.

Reply:
Sorry? I didn't understand

Reply:
`info.set_primary_lock(lock.primary)` moves `lock.primary` to `info`. No need to use `std::mem::replace`.
![image](https://user-images.githubusercontent.com/14819777/65400339-32781500-ddf4-11e9-8c70-0f1f13c7361a.png)

Reply:
The `lock` comes from `Some(ref mut lock)` in the `match`'s arm, so as a reference it's not possible to move it's `primary` field directly. If matches `Some(mut lock)` instead of ref, it's not allowed to use `if lock.ts == self.start_ts` in the match arm (match guard). The compiler tells me that I can still do this by adding a `#![feature(...)]`, but being not sure if it causes potential trouble, I come up with this way to use mem::replace to solve the problem.

Reply:
My fault.🤣